cloud/openstack/os_stack.py

```diff
@@ -0,0 +1,258 @@
+        type: dict
+        sample: "{'OS::project_id': '7f6a3a3e01164a4eb4eecb2ab7742101',
+                    'OS::stack_id': '97a3f543-8136-4570-920e-fd7605c989d6',
+                    'OS::stack_name': 'test-stack',
+                    'stack_status': 'CREATE_COMPLETE',
+                    'stack_status_reason': 'Stack CREATE completed successfully',
+                    'status': 'COMPLETE',
+                    'template_description': 'HOT template to create a new instance and networks',
+                    'timeout_mins': 60,
+                    'updated_time': null}"
+'''
+
+def _create_stack(module, stack, cloud):
+    try:
+        stack = cloud.create_stack(module.params['name'],
+                                       template_file=module.params['template'],
+                                       environment_files=module.params['environment'],
+                                       timeout=module.params['timeout'],
+                                       wait=True,
+                                       rollback=module.params['rollback'],
+                                       **module.params['parameters'])
+
+        stack = cloud.get_stack(stack.id, None)
+        if stack.stack_status == 'CREATE_COMPLETE':
+            return stack
+        else:
+            return False
+            module.fail_json(msg = "Failure in creating stack: ".format(stack))
+    except shade.OpenStackCloudException as e:
+        module.fail_json(msg=str(e))
+
+def _update_stack(module, stack, cloud):
+    try:
+        stack = cloud.update_stack(
+            module.params['name'],
+            template_file=module.params['template'],
+            environment_files=module.params['environment'],
+            timeout=module.params['timeout'],
+            rollback=module.params['rollback'],
+            wait=module.params['wait'])
+
+        if stack['stack_status'] == 'UPDATE_COMPLETE':
+            return stack
+        else:
+            module.fail_json(msg = "Failure in updating stack: %s" %
+                             stack['stack_status_reason'])
+    except shade.OpenStackCloudException as e:
+        module.fail_json(msg=str(e))
+
+def _system_state_change(module, stack, cloud):
+    state = module.params['state']
+    if state == 'present':
+        if not stack:
+            return True
+    if state == 'absent' and stack:
+        return True
+    return False
+
+def main():
+
+    argument_spec = openstack_full_argument_spec(
+        name=dict(required=True),
+        template=dict(default=None),
+        environment=dict(default=None, type='list'),
+        parameters=dict(default={}, type='dict'),
+        rollback=dict(default=False, type='bool'),
+        timeout=dict(default=3600, type='int'),
+        state=dict(default='present', choices=['absent', 'present']),
+    )
+
+    module_kwargs = openstack_module_kwargs()
+    module = AnsibleModule(argument_spec,
+                           supports_check_mode=True,
```

Comment:
Excellent catch by @resmo. You define a _system_state_change() function, yet it is never used.

needs_revision